### PR TITLE
adds datamodel for channel and update visualisation

### PR DESCRIPTION
**Pull request by invincibleJai**
**Fixes**: 
- https://issues.redhat.com/browse/ODC-4164
- https://issues.redhat.com/browse/ODC-4165
- https://issues.redhat.com/browse/ODC-4169

**Analysis / Root cause**: 
In topology we don't visualise Channels / Subscription associated relation with Sources and kn service

**Solution Description**: 
- Fetch Channels CRDs based on label and and form Datamodel required for visualising Channel  / Subscription associated relation
- Extending Shape for Channel
- Extending connector for Subscription
- Show sidebar for channel  
- Show sidebar for Subscription connector
- Update sidebar for EventSources
- Action menus for Channel Node/Subscription Edge - Move Subscription apart from custom one's
- Gestures for for eventSource Connector can be dropped to Channel as well
- Gestures for for Subscription  Connector can be dropped to Other kn Svc as well

**Screen shots / Gifs for design review**: 

- Sidebar for Channel
image

- Sidebar for Subscription connector
image

- Sidebar for eventSource while sinking to channel
image

- gif showing some of gestures and action menu

Jun-26-2020 09-46-13

@openshift/team-devconsole-ux

**Unit test coverage report**: 
image

**Test setup:**
- Create channel/sources/subscription and kn svc, can follow https://github.com/matzew/knative-eventing-samples/tree/master/02-source_channel

**Browser conformance**: 

- [x] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge

### CODE REVIEW

File: frontend/packages/knative-plugin/src/topology/knative-topology-utils.ts [hunk 1]
```diff
@@ -269,6 +274,74 @@ const createKnativeDeploymentItems = (
...
+export const createPubSubDataItems = (
+  resource: K8sResourceKind,
+  resources: TopologyDataResources,
+) => {
+  const {
+    kind: resKind,
+    metadata: { name },
+    spec,
+  } = resource;
+  const subscriptionData = _.get(resources, 'eventingsubscription.data', []);
```
Comment (jeff-phillips-18) [inline frontend/packages/knative-plugin/src/topology/knative-topology-utils.ts:302]: We should be avoiding the use of `_.get`

File: frontend/packages/knative-plugin/src/topology/components/nodes/EventingPubSubNode.tsx [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+  const resourceObj = getTopologyResourceObject(element.getData());
+  const resourceModel = modelFor(referenceFor(resourceObj));
+  const editAccess = useAccessReview({
+    group: resourceModel.apiGroup,
+    verb: 'patch',
+    resource: resourceModel.plural,
+    name: resourceObj.metadata.name,
+    namespace: resourceObj.metadata.namespace,
+  });
```
Comment (karthikjeeyar) [inline frontend/packages/knative-plugin/src/topology/components/nodes/EventingPubSubNode.tsx:69]: withEditReviewAccess HOC is used in componentFactory, so this component props will aleady have `editAccess` in its props, so we can remove this.

File: frontend/packages/knative-plugin/src/topology/components/nodes/EventingPubSubNode.tsx [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+  React.useLayoutEffect(() => {
+    if (editAccess) {
+      if (hover) {
+        onShowCreateConnector && onShowCreateConnector();
+      } else {
+        onHideCreateConnector && onHideCreateConnector();
+      }
+    }
+  }, [editAccess, hover, onShowCreateConnector, onHideCreateConnector]);
```
Comment (karthikjeeyar) [inline frontend/packages/knative-plugin/src/topology/components/nodes/EventingPubSubNode.tsx:93]: We can remove this create connector as this PR doesn't deal with createc connector HOC, WDYT?

File: frontend/packages/knative-plugin/src/topology/components/nodes/EventingPubSubNode.tsx [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+export type EventingPubSubNodeProps = {
+  element: Node;
+  canDrop?: boolean;
+  dropTarget?: boolean;
+  edgeDragging?: boolean;
+} & WithSelectionProps &
+  WithDndDropProps &
+  WithContextMenuProps &
+  WithCreateConnectorProps;
```
Comment (karthikjeeyar) [inline frontend/packages/knative-plugin/src/topology/components/nodes/EventingPubSubNode.tsx:43]: nit: can be removed.

File: frontend/packages/knative-plugin/src/topology/knative-topology-utils.ts [hunk 5]
```diff
@@ -269,6 +274,74 @@ const createKnativeDeploymentItems = (
...
+export const getKnativeEventSources = (resources: TopologyDataResources): K8sResourceKind[] => {
+  const evenSourceProps = getDynamicEventSourcesModelRefs();
+  return evenSourceProps.reduce((acc, currProp) => {
+    const currPropResource = resources[currProp]?.data ?? [];
+    return [...acc, ...currPropResource];
+  }, []);
+};
+
+export const getKnativeChannelResources = (resources: TopologyDataResources): K8sResourceKind[] => {
+  const evenSourceProps = getDynamicChannelModelRefs();
```
Comment (karthikjeeyar) [inline frontend/packages/knative-plugin/src/topology/knative-topology-utils.ts:286]: nit: 
```suggestion
  const channelProps = getDynamicChannelModelRefs();
```

File: frontend/packages/knative-plugin/src/topology/knative-topology-utils.ts [hunk 6]
```diff
@@ -269,6 +274,74 @@ const createKnativeDeploymentItems = (
   };
 };
 
+export const getKnativeEventSources = (resources: TopologyDataResources): K8sResourceKind[] => {
+  const evenSourceProps = getDynamicEventSourcesModelRefs();
```
Comment (karthikjeeyar) [inline frontend/packages/knative-plugin/src/topology/knative-topology-utils.ts:278]:
```suggestion
  const eventSourceProps = getDynamicEventSourcesModelRefs();
```

File: frontend/packages/knative-plugin/src/topology/knative-topology-utils.ts [hunk 7]
```diff
@@ -318,14 +391,15 @@ export const getKnativeTopologyNodeItems = (
...
+  const sinkTarget = _.get(resource, 'spec.sink.ref', null) || _.get(resource, 'spec.sink', null);
   const edges = [];
-  if (sinkSvc && sinkSvc.kind === knServiceModel.kind) {
+  if (sinkTarget) {
     _.forEach(data, (res) => {
-      const resname = _.get(res, ['metadata', 'name']);
-      const resUid = _.get(res, ['metadata', 'uid']);
-      if (resname === sinkSvc.name) {
+      const {
+        metadata: { uid: resUid, name: resname },
```
Comment (karthikjeeyar) [inline frontend/packages/knative-plugin/src/topology/knative-topology-utils.ts:400]: nit
```suggestion
        metadata: { uid: resUid, name: resName },
```

File: frontend/packages/knative-plugin/src/topology/knative-topology-utils.ts [hunk 8]
```diff
@@ -338,6 +412,43 @@ export const getEventTopologyEdgeItems = (resource: K8sResourceKind, { data }):
...
+  } = resource;
+  const { eventingsubscription, ksservices } = resources;
+  const edges = [];
+  _.forEach(eventingsubscription?.data, (subRes) => {
+    const channelData = _.get(subRes, ['spec', 'channel']);
+    if (name === channelData?.name && ksservices) {
+      const svcData = _.get(subRes, ['spec', 'subscriber', 'ref']);
+      _.forEach(ksservices?.data, (res) => {
+        const {
+          metadata: { uid: resUid, name: resname },
```
Comment (karthikjeeyar) [inline frontend/packages/knative-plugin/src/topology/knative-topology-utils.ts:430]:
```suggestion
          metadata: { uid: resUid, name: resName },
```

File: frontend/packages/knative-plugin/src/topology/components/nodes/EventingPubSubNode.tsx [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+  canDrop,
+  dropTarget,
+  selected,
+  onSelect,
+  onContextMenu,
+  contextMenuOpen,
+  dndDropRef,
+  edgeDragging,
+  onHideCreateConnector,
+  onShowCreateConnector,
```
Comment (karthikjeeyar) [inline frontend/packages/knative-plugin/src/topology/components/nodes/EventingPubSubNode.tsx:56]: can be removed as well

File: frontend/packages/knative-plugin/src/topology/knative-topology-utils.ts [hunk 10]
```diff
@@ -505,3 +652,36 @@ export const createSinkConnection = (
...
+  const updatePayload = {
+    ...subscriptionObj,
+    spec: { ...subscriptionObj.spec, subscriber: { ...sink } },
+  };
+
+  return k8sUpdate(modelFor(referenceFor(subscriptionObj)), updatePayload);
+};
+
+export const createSinkPubSubConnection = (
+  connector: any,
```
Comment (karthikjeeyar) [inline frontend/packages/knative-plugin/src/topology/knative-topology-utils.ts:677]: can we pass sourceNode here and type it as `Node` and extract resources from the sourceNode?

**Comment by jeff-phillips-18** [comment]
Tested locally. Works great.

File: frontend/packages/knative-plugin/src/topology/knative-topology-utils.ts (hunk 10 above)
Comment (invincibleJai) [reply frontend/packages/knative-plugin/src/topology/knative-topology-utils.ts:672]: This util will enable change of subscriber from one knSvc to another through drag and drop or action menus so used connector resource. As connector here is subscription and it has all the info i.e channel and subscriber. Source node is channel that doesn't have these relationships and one channel can go to multiple subscriptions so even if we add it in resources object, still would need connector in context to map.

**Pull request merged**